src/burst_buffer/burst_buffer.lua

```diff
@@ -1,39 +1,722 @@
+	if rc[3] ~= 0 then
+		return false, result
+	end
+	return true, result
+end
+
+-- find_dw_directives will search for any requested directive lines in the job
+-- script and return them in a table.
+function find_dw_directives(job_script)
+	local dw_directives = {}
+	local idx = 1
+	local bb
+	local line
+
+	io.input(job_script)
+	local content = io.read("*all")
+
+	local mline = "^#" .. DIRECTIVE .. " [^\n]+"
+	for line in content:gmatch("[^\n]+") do
+		bb = line:match(mline)
+		if bb ~= nil then
+			dw_directives[idx] = bb
+			idx = idx + 1
+		end
+	end
+	return dw_directives
+end
+
+-- make_workflow populates a workflow resource and submits it to DWS.
+-- On success this returns true and a nil error message.
+-- On failure this returns false and an error message.
+function make_workflow(workflow, job_script, jobid, userid, groupid)
+	local dwd = find_dw_directives(job_script)
+	workflow:initialize(WLMID_PLACEHOLDER, jobid, userid, groupid, dwd)
+
+	local yaml_name = os.tmpname()
+	local done, err = workflow:save(yaml_name)
+	if done == false then
+		return done, err
+	end
+
+	done, err = workflow:apply(yaml_name)
+	os.remove(yaml_name)
+	if done == false then
+		return done, err
+	end
+
+	return true, nil
+end
+
+-- make_workflow_name constructs a workflow name with the given job ID.
+function make_workflow_name(job_id)
+	return "bb" .. job_id
+end
+
+--[[
+--slurm_bb_job_process
+--
+--WARNING: This function is called synchronously from slurmctld and must
+--return quickly.
+--
+--This function is called on job submission.
+--
+--Slurm has just created a job ID and created the job script file.  We're now
+--asked to validate the script.  In DWS-speak, this means we have to define a
+--Workflow resource and submit it to DWS where the validation steps occur.  If
+--the validation steps succeed then a Workflow resource will exist and will be
+--in DWS's "Proposal" state.
+--
+--Slurm does not give us the job ID, user ID, or group ID at this time, so
+--placeholder values will be used.  Slurm will give us those values when it
+--asks us to transition to setup state and we'll patch the Workflow resource
+--at that time.
+--
+--We do not wait for the proposal state to transition to "Completed". If we did
+--not get an error on the initial apply of the resource then we know it passed
+--all validation steps. Any errors which may occur later to prevent the state's
+--transition to "Completed" will be caught when we attempt to transition it to
+--"Setup" state.
+--
+--If the DWS validation steps fail then the Workflow resource will not exist
+--and there will be no need for cleanup.
+--
+--If this function returns an error, the job is rejected and the second return
+--value (if given) is printed where salloc, sbatch, or srun was called.
+--]]
+function slurm_bb_job_process(job_script)
+	slurm.log_info("%s: slurm_bb_job_process(). job_script=%s",
+		lua_script_name, job_script)
+
+	-- Note: In this version of Slurm we do not have the job ID in this
```

Comment:
Specify which version of Slurm "this version" is. You should also be specific about which version of Slurm introduces the job id in slurm_bb_job_process.